homeassistant/components/calendar/outlook.py

```diff
@@ -0,0 +1,512 @@
+    vol.Required(CONF_DEVICE_ID): cv.string,
+    vol.Optional(CONF_TRACK): cv.boolean,
+    vol.Optional(CONF_SEARCH): vol.Any(cv.string, None),
+    vol.Optional(CONF_OFFSET): cv.string,
+})
+
+DEVICE_SCHEMA = vol.Schema({
+    vol.Required(CONF_CAL_ID): cv.string,
+    vol.Required(CONF_ENTITIES, None):
+        vol.All(cv.ensure_list, [_SINGLE_CALSEARCH_CONFIG]),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Setup the Outlook platform."""
+    # Assign configuration variables.
+    # The configuration check takes care they are present.
+    client_id = config[CONF_CLIENT_ID]
+    client_secret = config[CONF_CLIENT_SECRET]
+
+    oauth = OutlookAuthHelper(hass, client_id, client_secret)
+    if oauth.config_is_valid is False:
+        request_app_setup(hass, config, oauth, add_devices)
+        return False
+
+    service = OutlookService(hass, oauth)
+    add_devices(service.calendars)
+
+
+def config_from_file(filename, config=None):
+    """Small configuration file management function."""
+    if config:
+        # We"re writing configuration
+        try:
+            with open(filename, 'w') as fdesc:
+                fdesc.write(json.dumps(config))
+        except IOError as error:
+            _LOGGER.error("Saving config file failed: %s", error)
+            return False
+        return config
+    else:
+        # We"re reading config
+        if os.path.isfile(filename):
+            try:
+                with open(filename, 'r') as fdesc:
+                    return json.loads(fdesc.read())
+            except IOError as error:
+                _LOGGER.error("Reading config file failed: %s", error)
+                # This won"t work yet
+                return False
+        else:
+            return {}
+
+
+def request_app_setup(hass, config, oauth, add_devices):
+    """Request application setup."""
+    start_url = '{}{}'.format(hass.config.api.base_url,
+                              OUTLOOK_AUTH_START)
+    callback_url = '{}{}'.format(hass.config.api.base_url,
+                                 OUTLOOK_AUTH_CALLBACK_PATH)
+
+    outlook_auth_start_url = oauth.get_signin_url()
+
+    print('Outlook.com Authenticate:{}'.format(outlook_auth_start_url))
+
+    hass.http.register_redirect(OUTLOOK_AUTH_START,
+                                outlook_auth_start_url)
+    hass.http.register_view(
+        OutlookAuthCallbackView(config, add_devices, oauth))
+
+    persistent_notification = loader.get_component('persistent_notification')
+    persistent_notification.create(
+        hass, 'In order to authorize Home-Assistant<br/>'
+              'to view your calendars<br />'
+              'you must visit: <a href="{}" target="_blank">{}</a><br />'
+              'Callback url: {}'
+              ''.format(start_url, start_url, callback_url),
+        title=NOTIFICATION_TITLE, notification_id=NOTIFICATION_ID
+    )
+
+
+class OutlookService():
```

Comment:
This is too much protocol specific information. Please extract this into a 3rd party lib.